realm/realm-library/src/main/cpp/java_binding_context.cpp

```diff
@@ -17,18 +17,40 @@
 #include "java_binding_context.hpp"
 #include "jni_util/java_method.hpp"
 
+#include "util.hpp"
+
 using namespace realm;
 using namespace realm::_impl;
 using namespace realm::jni_util;
 
-void JavaBindingContext::changes_available()
+void JavaBindingContext::before_notify()
 {
+    if (JniUtils::get_env()->ExceptionCheck()) {
+        return;
```

Comment:
Returning to Java land immediately without touching JNI can be close to impossible. My point is that clearing the exception and throw a C++ exception might be safer. With a C++ exception, you will end at `CATCH_STD` and the border immediately instead of relying on unwinding the stack using `return`.